src/sinks/util/adaptive_concurrency/mod.rs

```diff
@@ -21,24 +21,32 @@ pub(self) fn instant_now() -> std::time::Instant {
 // The defaults for these values were chosen after running several
 // simulations on a test service that had various responses to load. The
 // values are the best balances found between competing outcomes.
-#[derive(Clone, Copy, Debug, Derivative, Deserialize, Serialize)]
-#[derivative(Default)]
+#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
 pub struct AdaptiveConcurrencySettings {
     // This value maintained high concurrency without holding it too
     // high under adverse conditions.
-    #[serde(default)]
-    #[derivative(Default(value = "0.9"))]
+    #[serde(default = "default_decrease_ratio")]
```

Comment:
Ah I see. Could this also be resolved by leaving the `derivative(Default)`s here, removing the field level `serde(default)`s and adding `serde(default)` to the struct? Mostly just for my own edification.

Reply:
I don't actually know. I'll play around with it in the playground.

Reply:
I couldn't get any combination of defaults to produce this effect except for having explicit values in the `serde(default)` clauses.

Reply:
Hmm, I was curious because I thought it should work and this _seems_ to work:

```rust
use derivative::Derivative;
use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, Derivative, Deserialize, Serialize)]
#[derivative(Default)]
#[serde(default)]
struct AdaptiveConcurrencySettings {
    // This value maintained high concurrency without holding it too
    // high under adverse conditions.
    #[derivative(Default(value = "0.9"))]
    decrease_ratio: f64,

    // This value achieved the best balance between quick response and
    // stability.
    #[derivative(Default(value = "0.7"))]
    ewma_alpha: f64,

    // This value avoided changing concurrency too aggressively when
    // there is fluctuation in the RTT measurements.
    #[derivative(Default(value = "0.05"))]
    rtt_threshold_ratio: f64,
}

fn main() {
    let settings: AdaptiveConcurrencySettings = serde_json::from_str("{}").unwrap();
    dbg!(&settings);
}
```

I may be missing something though.

Reply:
Ha, you're right, and that's sooo much better, thanks.